### PR TITLE
all: add a go-cloud useragent to GCP requests

### PR DESCRIPTION
**Pull request by vangent**
This is best-effort for gRPC; it's added in the default `Dial` function we provide for `runtimevar`, but the caller can always build their own.

For HTTP, we can guarantee it by wrapping the user-provided `HTTPClient`.

Feedback on the actual string I chose is welcome.

### CODE REVIEW

**Comment by vangent** [comment]
Note: tests will fail, as I need to re-record a bunch of replay files with the new User-Agent. I will do that after approval to avoid horrible conflicts.

```
replay.go:106: mismatched HTTP header "User-Agent" header at request #0; got "google-api-go-client/0.5 go-cloud/0.1" want "google-api-go-client/0.5"
```

File: internal/useragent/useragent.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+// Package useragent includes constants and utilitiesfor setting the User-Agent
+// for Go Cloud connections to GCP.
+package useragent // import "gocloud.dev/internal/useragent"
+
+import (
+	"net/http"
+)
+
+// GoCloudUserAgent is the User-Agent to set for Go Cloud requests to GCP.
+const GoCloudUserAgent = "go-cloud/0.1"
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/useragent/useragent.go]: I have no plan. What do the Veneer and other GCP APIs do?

BTW, we don't necessarily have to keep it in sync, unless we want to track usage by version.

File: internal/useragent/useragent.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func (t *userAgentTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Clone the request to avoid mutating it.
+	newReq := *req
+	newReq.Header = make(http.Header)
+	for k, vv := range req.Header {
+		newReq.Header[k] = vv
+	}
+	// Append, don't overwrite.
```
Comment (zombiezen) [inline internal/useragent/useragent.go]: I originally thought this was referring to the memory layout, but I see it's referring the content of the string. I suggest updating the comment.

```suggestion
	// Append to the User-Agent string to preserve other information.
```

Should this check for empty user-agent and avoid prepending the space?

**Comment by vangent** [comment]
> Shouldn't server also wrap the connection? It establishes the OpenCensus exporters.

I am not sure what you mean by this, can you clarify? Maybe appropriate to do in a separate PR...?

**Comment by zombiezen** [comment]
`server` takes in `[]option.ClientOption`:

https://github.com/google/go-cloud/blob/3913b382077c4f0c5d19d91c550490abec5a1419/server/sdserver/server.go#L52-L57

Seems like that should include `option.WithUserAgent` if the scope of this PR is to set the User-Agent string everywhere.

**Comment by vangent** [comment]
> server takes in []option.ClientOption

Ah, I see. Done.

**Pull request merged**